internal/datasource/http_auth.go

```diff
@@ -0,0 +1,299 @@
+package datasource
+
+import (
+	"bytes"
+	"context"
+	"crypto/md5"
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/hex"
+	"fmt"
+	"net/http"
+	"slices"
+	"strings"
+	"sync/atomic"
+)
+
+type HTTPAuthMethod int
+
+const (
+	AuthBasic HTTPAuthMethod = iota
+	AuthBearer
+	AuthDigest
+)
+
+// HTTPAuthentication holds the information needed for general HTTP Authentication support.
```

Comment:
Does this have OSV-Scanner specific code in it? It sounds like something that'd be available in third-party libraries.
If so maybe we can add to the comment why we're using this instead of an existing lib.

Reply:
I actually struggled to find third-party libraries that did this HTTP authentication for clients.

But regardless, this `HTTPAuthentication` struct was made to be configurable enough to be able to copy the how the native `mvn`/`npm` tooling approaches authentication. We could mention something to that effect in the comment.